### PR TITLE
[6.0-rc2] Query: Translate ToList over grouping element

### PR DESCRIPTION
**Pull request by smitpatel**
Part 2 & 3 of #26046

**Description**
- Query with GroupBy as final operator throws unfriendly error message
- Query which specifies ToList over grouping element throws translation error or generate incorrect results

**Customer Impact**
Customer will get exception thrown or incorrect results

**How found**
Manual testing done by @ajcvickers 

**Test coverage**
Added regression tests for scenario and similar cases.

**Regression?**
- Query with GroupBy as final operator throwing unfriendly error message is a regression (we used to throw better message)
- Query with ToList over grouping element is a new feature in 6.0

**Risk**
- Low. Code change is minimal and we have many tests going over same code path.

### CODE REVIEW

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 1]
```diff
@@ -1126,15 +1127,15 @@ public void ApplyPredicate(SqlExpression sqlExpression)
         ///     Applies grouping from given key selector.
         /// </summary>
         /// <param name="keySelector"> An key selector expression for the GROUP BY. </param>
-        public Expression ApplyGrouping(Expression keySelector)
+        public void ApplyGrouping(Expression keySelector)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1130]: This API was changed in 6.0, now it goes back to what it was in 5.0

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -1163,19 +1164,84 @@ public Expression ApplyGrouping(Expression keySelector)
...
             }
-
-            return keySelector;
         }
 
         /// <summary>
-        ///     Clears existing group by terms.
+        ///     Applies grouping from given key selector and generate <see cref="GroupByShaperExpression"/> to shape results.
         /// </summary>
-        public void ClearGroupBy()
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1173]: This was new API in 6.0

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 3]
```diff
@@ -1163,19 +1164,84 @@ public Expression ApplyGrouping(Expression keySelector)
...
+                keySelectorToAdd = sqlExpressionFactory.ApplyDefaultTypeMapping(sqlExpressionFactory.Constant(1));
+            }
+
+            var groupByTerms = new List<SqlExpression>();
+            var groupByAliases = new List<string?>();
+            PopulateGroupByTerms(keySelectorToAdd, groupByTerms, groupByAliases, "Key");
+
+            if (groupByTerms.Any(e => e is SqlConstantExpression || e is SqlParameterExpression || e is ScalarSubqueryExpression))
+            {
+                // EmptyKey will always hit this path.
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1196]: What's `EmptyKey`?

Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1196]: the variable defined above.

**Review by AndriySvyryd — APPROVED** [review]
🐑 🇮🇹

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1196]: Oh, you mean `emptyKey`

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1196]: :trollface:

**Comment by smitpatel** [comment]
@Pilchie @dotnet/efteam - This needs to be merged to rc2 branch. Need approval.

This couldn't finish in time because helix/windows builds are timing out.

**Comment by ajcvickers** [comment]
@smitpatel Please fill out the servicing template above.

**Comment by smitpatel** [comment]
Filled in template.

**Comment by Pilchie** [comment]
This introduces RC1 -> RC2 breaking API changes, right?  Do we have a communication plan for those?

**Comment by smitpatel** [comment]
The API being changed above are public mainly for providers. No customer should be using it in normal course of action. Further, even with the change, most of the work is done by relational provider itself without providers having to write a code which uses that API. There are only handful of providers which would have upgraded upto rc1 packages at this point. Perhaps we can just tag in the post to communicate?

**Comment by ajcvickers** [comment]
@smitpatel Can we verify that our major external providers for MySQL (@lauxjpn) and ProsgreSQL (@roji) will not break going from RC1 to RC2? This well help ensure people can quickly adopt RC2 when it goes out.

**Comment by smitpatel** [comment]
MySql doesn't have rc1 package yet.
My github search skills indicated that efcore.pg is not using any of above 2 APIs. @roji can provider better confirmation.

**Comment by Pilchie** [comment]
Okay, I'll approve based on that. Can we warn the MySQL author that this is coming in case they make an RC1 package between now and when RC2 releases?

**Comment by roji** [comment]
All OK from my side (efcore.pg).

**Comment by smitpatel** [comment]
@lauxjpn - If you are releasing package for rc1, please let me know if you need to use any of the API above. I can provide a work-around. (perhaps add them in future release if they are needed by providers).

**Comment by lauxjpn** [comment]
> @lauxjpn - If you are releasing package for rc1, please let me know if you need to use any of the API above. I can provide a work-around. (perhaps add them in future release if they are needed by providers).

@smitpatel Thanks, should not be an issue for us.

**Pull request merged**